rust/kernel/xarray.rs

```diff
@@ -0,0 +1,290 @@
+///   xarr.set(0, Box::try_new(0)?);
+///   xarr.get_shared(0, |x| assert_eq!(*x.unwrap(), 0));
+/// }
+/// {
+///   let mut xarr = xarr.as_mut();
+///   let mut value = xarr.get_mutable(0).unwrap();
+///   let mut value = value.deref_mut().as_mut();
+///   assert_eq!(value, &mut 0);
+///   *value = 1;
+/// }
+/// {
+///   let xarr = xarr.as_ref();
+///   xarr.get_shared(0, |x| assert_eq!(*x.unwrap(), 1));
+/// }
+///
+/// # Ok::<(), Error>(())
+/// ```
+///
+pub struct XArray<T: ForeignOwnable> {
+    xa: Opaque<bindings::xarray>,
+    _p: PhantomData<T>,
+    _q: PhantomPinned,
+}
+
+impl<T: ForeignOwnable> XArray<T> {
+    /// Creates a new `XArray` with the given flags.
+    pub fn new(flags: Flags) -> XArray<T> {
+        let xa = Opaque::uninit();
+
+        // SAFETY: We have just created `xa`. This data structure does not require
+        // pinning.
+        unsafe { bindings::xa_init_flags(xa.get(), flags) };
+
+        // INVARIANT: Initialize the `XArray` with a valid `xa`.
+        XArray {
+            xa,
+            _p: PhantomData,
+            _q: PhantomPinned,
+        }
+    }
+
+    /// Replaces an entry with a new value, returning the old value (if any).
+    pub fn replace(self: Pin<&Self>, index: usize, value: T) -> Result<Option<T>> {
+        let new = value.into_foreign();
+        // SAFETY: `new` just came from into_foreign(), and we dismiss this guard if
+        // the xa_store operation succeeds and takes ownership of the pointer.
+        let guard = ScopeGuard::new(|| unsafe {
+            T::from_foreign(new);
+        });
+
+        // SAFETY: `self.xa` is always valid by the type invariant, and we are storing
+        // a `T::into_foreign()` result which upholds the later invariants.
+        let old = unsafe {
+            bindings::xa_store(
+                self.xa.get(),
+                index.try_into()?,
+                new as *mut _,
+                bindings::GFP_KERNEL,
+            )
+        };
+
+        let ret = unsafe { bindings::xa_err(old) };
+        if ret != 0 {
+            Err(Error::from_errno(ret))
+        } else if old.is_null() {
+            guard.dismiss();
+            Ok(None)
+        } else {
+            guard.dismiss();
+            // SAFETY: The old value must have been stored by either this function or
+            // `alloc_limits_opt`, both of which ensure non-NULL entries are valid
+            // ForeignOwnable pointers.
+            Ok(Some(unsafe { T::from_foreign(old) }))
+        }
+    }
+
+    /// Replaces an entry with a new value, dropping the old value (if any).
+    pub fn set(self: Pin<&Self>, index: usize, value: T) -> Result {
+        self.replace(index, value)?;
+        Ok(())
+    }
+
+    /// Looks up a reference to an entry in the array, cloning it
+    /// and returning the cloned value to the user.
+    ///
+    pub fn get(self: Pin<&Self>, index: u64) -> Option<T::Borrowed<'_>>
+    where
+        T: Clone,
+    {
```

Comment:
Hmm, we might need a different trait than `Clone` here. For example, an `ArcBorrow` is not `Clone`, and the type we want to return is `Arc<T>` rather than `ArcBorrow`.